### PR TITLE
Add support to check if the selected org is Tanzu Application platform enabled

### PR DESCRIPTION
**Pull request by prkalle**
### What this PR does / why we need it

This PR adds support to check if the selected org is Tanzu Application Platform enabled

Summary of changes:

- Added support to show warning to users if the CSP token received through login doesn't contain the necessary TAP scopes. The TAP scopes are fetched from the CLI central configuration file and can be modified without releasing new CLI version.

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR

Used the local test registry to populate the central configuration and tested with organization with Tanzu Application Platform enabled. It was success

```
# start  the local test central repository
❯  make start-test-central-repo 

# update the CLI plugin default source to use local test cental repository
❯ tanzu plugin source update default -u localhost:9876/tanzu-cli/plugins/central:small

❯ ./bin/tanzu context create testTAPOrgCtx --type tanzu --staging --endpoint https://api.tanzu-dev.cloud.vmware.com
[i] Opening the browser window to complete the login
Log in by visiting this link:

    https://console-stg.cloud.vmware.com/csp/gateway/discovery?client_id=tanzu-cli-client-id&code_challenge=tdCdvLqhx8u0K2rjKx81cavc9r7TpDKKQh7vKbANNmU&code_challenge_method=S256&redirect_uri=http%3A%2F%2F127.0.0.1%3A64153%2Fcallback&response_type=code&state=56e24652ca96aec73024e9caf6dc05d6

    Optionally, paste your authorization code: [...]


[ok] Successfully logged into 'TestProjectSyncer' organization and created a tanzu context
[i] Checking for required plugins for context 'testTAPOrgCtx'...
[i] All required plugins are already installed and up-to-date
```
Used the local test registry to populate the central configuration and tested with organization WITHOUT Tanzu Application Platform enabled. It was success and the WARNING message is displayed stating the issue

```
❯ export TANZU_CLI_CLOUD_SERVICES_ORGANIZATION_ID=7753c388-914c-4085-88bc-1273f102c267
❯ ./bin/tanzu context create testNonTAPOrgCtx --type tanzu --staging --endpoint https://api.tanzu-dev.cloud.vmware.com
[i] Opening the browser window to complete the login
Log in by visiting this link:

    https://console-stg.cloud.vmware.com/csp/gateway/discovery?client_id=tanzu-cli-client-id&code_challenge=jZ4ayxsbXfKFt4p8QAKyr8GlpC1CDk2kBKgG_n2V40k&code_challenge_method=S256&orgId=7753c388-914c-4085-88bc-1273f102c267&redirect_uri=http%3A%2F%2F127.0.0.1%3A64263%2Fcallback&response_type=code&state=ddaf9f242d5b6784fd4910467b3d56d4

    Optionally, paste your authorization code: [...]


[ok] Successfully logged into '1_ARM_ESO_MAIN_FRESH' organization and created a tanzu context

[!] WARNING: While authenticated to organization '1_ARM_ESO_MAIN_FRESH', there are insufficient permissions to access
the Tanzu Application Platform service. Please ensure correct organization authentication and access permissions

[i] Checking for required plugins for context 'testNonTAPOrgCtx'...
[i] All required plugins are already installed and up-to-date
```

Verifed the TAP scope validations are skipped (and the warning message is not shown) during login/context creation by exporting the environment variable
```
❯ ./bin/tanzu login
[i] The tanzu cli essential plugins are outdated and are being updated now. The update may take a few seconds.
[i] Installing plugins from plugin group 'vmware-tanzucli/essentials:v9.9.9'
[i] Refreshing plugin inventory cache for "localhost:9876/tanzu-cli/plugins/central:small", this will take a few seconds.
[i] Installed plugin 'telemetry:v9.9.9' with target 'global' (from cache)

[i] Opening the browser window to complete the login
Log in by visiting this link:

    https://console.cloud.vmware.com/csp/gateway/discovery?client_id=tanzu-cli-client-id&code_challenge=N0o9d8wrYWmyPzpER3FvUj55HpBnvmpgzahqSNVgbYI&code_challenge_method=S256&redirect_uri=http%3A%2F%2F127.0.0.1%3A62923%2Fcallback&response_type=code&state=eee8dd54bb741e9e22dfa6284b4a50d8

    Optionally, paste your authorization code: [...]


[ok] Successfully logged into 'CMBU TMM' organization and created a tanzu context

[!] WARNING: While authenticated to organization 'CMBU TMM', there are insufficient permissions to access
the Tanzu Application Platform service. Please ensure correct organization authentication and access permissions

[i] Checking for required plugins for context 'CMBU_TMM'...
[i] All required plugins are already installed and up-to-date


❯ export TANZU_CLI_SKIP_TAP_SCOPES_VALIDATION_ON_TANZU_CONTEXT=true

❯ ./bin/tanzu login
[i] Opening the browser window to complete the login
Log in by visiting this link:

    https://console.cloud.vmware.com/csp/gateway/discovery?client_id=tanzu-cli-client-id&code_challenge=vahA5ae3GwlEWYYlJSs7lvlkP2YB69qBMDCPtYyLM80&code_challenge_method=S256&redirect_uri=http%3A%2F%2F127.0.0.1%3A63030%2Fcallback&response_type=code&state=ac7988dd35077dbd3682ab3069f1b425

    Optionally, paste your authorization code: [...]


[ok] Successfully logged into 'CMBU TMM' organization and created a tanzu context
[i] Checking for required plugins for context 'CMBU_TMM'...
[i] All required plugins are already installed and up-to-date
```

### Release note

```release-note
Add support to check if the selected org is Tanzu Application platform enabled
```

#### Special notes for your reviewer
We require to update the central configuration with the TAP scopes and publish to plugin repositories.

### CODE REVIEW

File: pkg/command/org_validation_helper.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+type tapScope struct {
+	Scope string `yaml:"scope" json:"scope"`
+}
+
+type tapScopesGetter func() ([]string, error)
+
+// validateTokenForTAPScopes validates if the claims contain required TAP scopes.
+func validateTokenForTAPScopes(claims *csp.Claims, scopesGetter tapScopesGetter) (bool, error) {
+	if scopesGetter == nil {
+		scopesGetter = getTAPScopesFromCentralConfig
```
Comment (vuil) [inline pkg/command/org_validation_helper.go:31]: nit: Do we need worry about insisting on a single set of scopes to be used by all tanzu contexts? 
at minimum should we provide a CLI option to disable the scope validation?

File: pkg/command/org_validation_helper.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the Tanzu Application Platform scopes from the central configuration
+	reader := centralconfig.NewCentralConfigReader(discoverySource)
+	var tapScopes []tapScope
+	err = reader.GetCentralConfigEntry(centralConfigTanzuApplicationPlatformScopesKey, &tapScopes)
+	if err != nil {
+		// If the key not found in the central config, it does not return an error because some central repositories
```
Comment (vuil) [inline pkg/command/org_validation_helper.go:68]: nit: is not found

File: pkg/command/org_validation_helper_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+func TestValidateTokenForTanzuScopes(t *testing.T) {
+	tests := []struct {
+		name         string
+		claims       *csp.Claims
+		scopesGetter tapScopesGetter
+		expectedBool bool
+		expectedErr  error
+	}{
+		{
+			name:         "NoTAPScopes_Success",
```
Comment (vuil) [inline pkg/command/org_validation_helper_test.go:22]: nit: proper english words in the name field?

File: pkg/command/org_validation_helper.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+type tapScopesGetter func() ([]string, error)
+
+// validateTokenForTAPScopes validates if the claims contain required TAP scopes.
+func validateTokenForTAPScopes(claims *csp.Claims, scopesGetter tapScopesGetter) (bool, error) {
+	if scopesGetter == nil {
+		scopesGetter = getTAPScopesFromCentralConfig
+	}
+	tapScopes, err := scopesGetter()
+	if err != nil {
+		log.V(7).Error(err, "error retrieving TAP scopes form the central config")
```
Comment (vuil) [inline pkg/command/org_validation_helper.go:35]: from the, same for next line

File: pkg/command/org_validation_helper_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+			expectedErr:  nil,
+		},
+		{
+			name:   "GetTAPScopesError_Failure",
+			claims: &csp.Claims{},
+			scopesGetter: func() ([]string, error) {
+				return nil, errors.New("error retrieving TAP scopes")
+			},
+			expectedBool: false,
+			expectedErr:  errors.New("error retrieving TAP scopes form the central config: error retrieving TAP scopes"),
```
Comment (vuil) [inline pkg/command/org_validation_helper_test.go:53]: from the

File: pkg/command/org_validation_helper.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+		return false, errors.Wrap(err, "error retrieving TAP scopes form the central config")
+	}
+
+	// validate only if we get the permissions/scopes from central configuration
+	if len(tapScopes) == 0 {
+		return true, nil
+	}
+
+	for _, reqTAPScope := range tapScopes {
```
Comment (vuil) [inline pkg/command/org_validation_helper.go:44]: nit: expand the variable or use a different name. Is `req` supposed to be for "required"
Based on the matching logic, as long as user's permissions intersects with the configured list of TAP scopes, even if just one item, the validation is successful.

On related note, we should doc or add a comment somewhere describing the behavior.

**Review by vuil — COMMENTED** [review]
a couple more minor things.

**Review by marckhouzam — APPROVED** [review]
Nice.  I like the use of the Central Config to allow to add or modify scopes whenever TAP evolves.  The tricky thing will be to inform the relevant people (who?) that when a new scope is added or a scope renamed, the Central Config must be updated.  But it is a quick fix if it is forgotten, and only a WARNING would show up.

Besides @vuil comments, LGTM.

File: pkg/command/org_validation_helper.go (hunk 4 above)
Comment (prkalle) [reply pkg/command/org_validation_helper.go:35]: Fixed. Thank you.

File: pkg/command/org_validation_helper_test.go (hunk 5 above)
Comment (prkalle) [reply pkg/command/org_validation_helper_test.go:53]: fixed

File: pkg/command/org_validation_helper_test.go (hunk 3 above)
Comment (prkalle) [reply pkg/command/org_validation_helper_test.go:22]: updated.

File: pkg/command/org_validation_helper.go (hunk 6 above)
Comment (prkalle) [reply pkg/command/org_validation_helper.go:44]: updated the variable name from `reqTAPScope` to `tapScope` and also updated the function doc comment describing the behavior.

File: pkg/command/context.go [hunk 7]
```diff
@@ -662,9 +662,27 @@ func globalTanzuLogin(c *configtypes.Context, generateContextNameFunc func(orgNa
...
+	valid, err := validateTokenForTAPScopes(claims, nil)
+	if err == nil && !valid {
+		logTanzuInvalidOrgWarningMessage(orgName)
+	}
 	return nil
 }
 
+func logTanzuInvalidOrgWarningMessage(orgName string) {
+	warnMsg := `WARNING: There seems to be an issue with the organization '%s' you are trying to log in to.
+The authentication token does not seem to contain the necessary permissions or scopes required for accessing the
```
[the remark this answers is not in the document]
Comment (prkalle) [reply pkg/command/context.go:676]: Thank you. Updated the text

[the remark this answers is not in the document]
Comment (prkalle) [reply pkg/command/context.go:676]: @vuil would like to get the feedback on the below text
`
WARNING: While authenticated to organization '%s', there are insufficient permissions to access the Tanzu Application Platform service.Please ensure correct organization authentication and access permissions
`

File: pkg/command/org_validation_helper.go (hunk 1 above)
Comment (prkalle) [reply pkg/command/org_validation_helper.go:31]: Thank you for pointing it out. Provided an environment variable (`TANZU_CLI_SKIP_TAP_SCOPES_VALIDATION_ON_TANZU_CONTEXT`) to skip the scope validation on tanzu contexts

File: pkg/command/context.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (vuil) [reply pkg/command/context.go]: That works for me. (Add space between . and 'Please...)

**Comment by prkalle** [comment]
> The tricky thing will be to inform the relevant people (who?) that when a new scope is added or a scope renamed, the Central Config must be updated.

Thanks @marckhouzam. I was thinking that central configuration would check (periodically) and would be updated  if it finds plugin inventory image is updated. Are you saying the user would see the warning for the time period between we post the updated inventory image(with new central config) and next automatic update that CLI would perform and download the latest central configuration? If so, yes I agree user would see this warning message till the next CLI update.

File: pkg/command/context.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (prkalle) [reply pkg/command/context.go]: updated.

**Review by vuil — APPROVED** [review]
LGTM, thanks for the updates.
Please mention the env var in the release notes section, and also add
doc-impact label.

**Pull request merged**